src/gpuarray_collectives_cuda_nccl.c

```diff
@@ -253,7 +253,7 @@ static int reduce(gpudata* src, size_t offsrc, gpudata* dest, size_t offdest,
   GA_EXIT_ON_ERROR(ctx, cuda_wait(src, CUDA_WAIT_READ));
   // sync: wait till a read/write has finished (out of concurrent kernels)
   if (rank == root)
-    GA_EXIT_ON_ERROR(ctx, cuda_wait(dest, CUDA_WAIT_WRITE));
+    GA_EXIT_ON_ERROR(ctx, cuda_wait(dest, CUDA_WAIT_ALL));
```

Comment:
[This](https://github.com/tsirif/libgpuarray/blob/feature/nccl/src/gpuarray_buffer_cuda.c#L612) is something I would like to discuss.

When we want to read a `gpudata`, we should wait for previously issued (earlier scheduled) writes to this thread to finish.
When we want to write to a `gpudata`, we should wait for previously issued (earlier scheduled) reads and writes to this thread to finish.

Also, in the case where all kernels run on the same stream, nothing is concurrent. So for the device there are no synchronization issues, because kernels are executed serially with the order they were issued. Synchronization issues occur in this case only between the stream and the host.

Do you agree on these? Shouldn't CUDA_WAIT_(TO)_WRITE also make the stream to wait for `a->wev` too? (And record to `a->wev` also when it finishes)